// @dcloudio/uni-cli-shared@3.0.0-alpha-3000020210521001/dist/json/pages.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizePagesJson = exports.parsePagesJsonOnce = exports.parsePagesJson = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const slash_1 = __importDefault(require("slash"));
const shared_1 = require("@vue/shared");
const uni_shared_1 = require("@dcloudio/uni-shared");
const json_1 = require("./json");
const parsePagesJson = (inputDir, platform, normalize = true) => {
    const jsonStr = fs_1.default.readFileSync(path_1.default.join(inputDir, 'pages.json'), 'utf8');
    if (normalize) {
        return normalizePagesJson(jsonStr, inputDir, platform);
    }
    return json_1.parseJson(jsonStr, true);
};
exports.parsePagesJson = parsePagesJson;
exports.parsePagesJsonOnce = uni_shared_1.once(exports.parsePagesJson);
function normalizePagesJson(jsonStr, inputDir, platform) {
    let pagesJson = {
        pages: [],
        globalStyle: {
            navigationBar: {},
        },
    };
    // preprocess
    try {
        pagesJson = json_1.parseJson(jsonStr, true);
    }
    catch (e) {
        console.error(`[vite] Error: pages.json parse failed.\n`, jsonStr, e);
    }
    // pages
    validatePages(pagesJson, jsonStr);
    // subpackages
    pagesJson.pages.push(...normalizeSubpackages(pagesJson.subPackages || pagesJson.subpackages));
    // pageStyle
    normalizePages(pagesJson.pages, inputDir, platform);
    // globalStyle
    pagesJson.globalStyle = normalizePageStyle(pagesJson.globalStyle, platform);
    // tabBar
    if (pagesJson.tabBar) {
        const tabBar = normalizeTabBar(pagesJson.tabBar);
        if (tabBar) {
            pagesJson.tabBar = tabBar;
        }
        else {
            delete pagesJson.tabBar;
        }
    }
    return pagesJson;
}
exports.normalizePagesJson = normalizePagesJson;
function validatePages(pagesJson, jsonStr) {
    if (!Array.isArray(pagesJson.pages)) {
        pagesJson.pages = [];
        console.error(`[uni-app] Error: pages.json->pages parse failed.\n`, jsonStr);
    }
    else if (!pagesJson.pages.length) {
        console.error(`[uni-app] Error: pages.json->pages must contain at least 1 page.\n`, jsonStr);
    }
}
function normalizePages(pages, _inputDir, platform) {
    return pages.filter((page) => {
        page.style = normalizePageStyle(page.style, platform);
        return true;
    });
}
function normalizeSubpackages(subpackages) {
    const pages = [];
    if (Array.isArray(subpackages)) {
        subpackages.forEach(({ root, pages: subPages }) => {
            if (root && subPages.length) {
                subPages.forEach((subPage) => {
                    subPage.path = slash_1.default(path_1.default.join(root, subPage.path));
                    pages.push(subPage);
                });
            }
        });
    }
    return pages;
}
function normalizePageStyle(pageStyle, platform) {
    if (pageStyle) {
        if (platform === 'h5') {
            Object.assign(pageStyle, pageStyle['app'] || {});
        }
        Object.assign(pageStyle, pageStyle[platform] || {});
        if (['h5', 'app'].includes(platform)) {
            pageStyle.navigationBar = normalizeNavigationBar(pageStyle);
        }
        return removePlatformStyle(pageStyle);
    }
    return { navigationBar: {} };
}
const navigationBarMaps = {
    navigationBarBackgroundColor: 'backgroundColor',
    navigationBarTextStyle: 'textStyle',
    navigationBarTitleText: 'titleText',
    navigationStyle: 'style',
    titleImage: 'titleImage',
    titlePenetrate: 'titlePenetrate',
};
function normalizeNavigationBar(pageStyle) {
    const navigationBar = Object.create(null);
    Object.keys(navigationBarMaps).forEach((name) => {
        if (shared_1.hasOwn(pageStyle, name)) {
            // @ts-ignore
            navigationBar[navigationBarMaps[name]] = pageStyle[name];
            delete pageStyle[name];
        }
    });
    const { titleNView } = pageStyle;
    if (shared_1.isPlainObject(titleNView)) {
        Object.assign(navigationBar, titleNView);
        delete pageStyle.titleNView;
    }
    if (!navigationBar.titleColor && shared_1.hasOwn(navigationBar, 'textStyle')) {
        navigationBar.titleColor =
            navigationBar.textStyle === 'black' ? '#000' : '#fff';
        delete navigationBar.textStyle;
    }
    if (pageStyle.navigationBarShadow &&
        pageStyle.navigationBarShadow.colorType) {
        navigationBar.shadowColorType = pageStyle.navigationBarShadow.colorType;
        delete pageStyle.navigationBarShadow;
    }
    if (shared_1.isArray(navigationBar.buttons)) {
        navigationBar.buttons = navigationBar.buttons.map((btn) => normalizeNavigationBarButton(btn, navigationBar.type, navigationBar.titleColor));
    }
    if (shared_1.isPlainObject(navigationBar.searchInput)) {
        navigationBar.searchInput = normalizeNavigationBarSearchInput(navigationBar.searchInput);
    }
    if (navigationBar.type === 'transparent') {
        navigationBar.coverage = navigationBar.coverage || '132px';
    }
    return navigationBar;
}
function normalizeNavigationBarButton(btn, type, titleColor) {
    btn.color = type === 'transparent' ? '#fff' : btn.color || titleColor;
    if (!btn.fontSize) {
        btn.fontSize =
            type === 'transparent' || (btn.text && /\\u/.test(btn.text))
                ? '22px'
                : '27px';
    }
    else if (/\d$/.test(btn.fontSize)) {
        btn.fontSize += 'px';
    }
    btn.text = btn.text || '';
    return btn;
}
function normalizeNavigationBarSearchInput(searchInput) {
    return Object.assign({
        autoFocus: false,
        align: 'center',
        color: '#000',
        backgroundColor: 'rgba(255,255,255,0.5)',
        borderRadius: '0px',
        placeholder: '',
        placeholderColor: '#CCCCCC',
        disabled: false,
    }, searchInput);
}
const DEFAULT_TAB_BAR = {
    position: 'bottom',
    color: '#999',
    selectedColor: '#007aff',
    borderStyle: 'black',
    blurEffect: 'none',
    fontSize: '10px',
    iconWidth: '24px',
    spacing: '3px',
    height: uni_shared_1.TABBAR_HEIGHT + 'px',
};
function normalizeTabBar(tabBar) {
    const { list, midButton } = tabBar;
    if (!list || !list.length) {
        return;
    }
    tabBar = shared_1.extend({}, DEFAULT_TAB_BAR, tabBar);
    const len = list.length;
    if (len % 2 === 0 && shared_1.isPlainObject(midButton)) {
        list.splice(Math.floor(len / 2), 0, shared_1.extend({
            type: 'midButton',
            width: '50px',
            height: '50px',
            iconWidth: '24px',
        }, midButton));
    }
    else {
        delete tabBar.midButton;
    }
    list.forEach((item) => {
        if (item.iconPath) {
            item.iconPath = normalizeFilepath(item.iconPath);
        }
        if (item.selectedIconPath) {
            item.selectedIconPath = normalizeFilepath(item.selectedIconPath);
        }
        if (item.type === 'midButton' && item.backgroundImage) {
            item.backgroundImage = normalizeFilepath(item.backgroundImage);
        }
    });
    tabBar.selectedIndex = 0;
    tabBar.shown = true;
    return tabBar;
}
const SCHEME_RE = /^([a-z-]+:)?\/\//i;
const DATA_RE = /^data:.*,.*/;
function normalizeFilepath(filepath) {
    if (!(SCHEME_RE.test(filepath) || DATA_RE.test(filepath)) &&
        filepath.indexOf('/') !== 0) {
        return '/' + filepath;
    }
    return filepath;
}
const platforms = ['h5', 'app', 'mp-', 'quickapp'];
function removePlatformStyle(pageStyle) {
    Object.keys(pageStyle).forEach((name) => {
        if (platforms.find((prefix) => name.startsWith(prefix))) {
            delete pageStyle[name];
        }
    });
    return pageStyle;
}
